### PR TITLE
[WIP DON'T MERGE] Playback tools

### PR DESCRIPTION
**Pull request by r2dliu**

### CODE REVIEW

File: Source/Core/Core/HW/EXI_DeviceSlippi.cpp [hunk 1]
```diff
@@ -21,8 +25,46 @@
...
+#include "Core/CoreTiming.h"
+#include "Core/HW/SystemTimers.h"
+#include "Core/State.h"
+
+// state diffs keyed by frameIndex;
+static std::map<int32_t, std::string> diffsByFrame;
+
+// diffs are processed async
+std::vector<std::future<std::pair<int32_t, std::string>>> futureDiffs;
+bool g_shouldJumpBack = false;
```
Comment (r2dliu) [inline Source/Core/Core/HW/EXI_DeviceSlippi.cpp]: doing this with a bunch of global/externs rn for testing, happy to use another pattern

File: Source/Core/Core/HW/EXI_DeviceSlippi.cpp [hunk 2]
```diff
@@ -812,3 +882,214 @@ bool CEXISlippi::IsPresent() const
...
+	open_vcdiff::VCDiffDecoder decoder;
+	std::tuple<int, size_t, std::vector<u8>> diffTuple;
+
+	while (true)
+	{
+		bool shouldSeek = g_inSlippiPlayback && hasRestartedReplay &&
+		                  (g_shouldJumpBack || g_shouldJumpForward || g_targetFrameNum != INT_MAX);
+		if (shouldSeek)
+		{
+			paused = (Core::GetState() == Core::CORE_PAUSE);
```
Comment (r2dliu) [inline Source/Core/Core/HW/EXI_DeviceSlippi.cpp]: should prob make a helper function for this chunk

File: Source/Core/Core/HW/EXI_DeviceSlippi.cpp [hunk 3]
```diff
@@ -812,3 +882,214 @@ bool CEXISlippi::IsPresent() const
...
+	ThreadPoolQueue pool(10);
+
+	open_vcdiff::VCDiffDecoder decoder;
+	std::tuple<int, size_t, std::vector<u8>> diffTuple;
+
+	while (true)
+	{
+		bool shouldSeek = g_inSlippiPlayback && hasRestartedReplay &&
+		                  (g_shouldJumpBack || g_shouldJumpForward || g_targetFrameNum != INT_MAX);
+		if (shouldSeek)
```
Comment (r2dliu) [inline Source/Core/Core/HW/EXI_DeviceSlippi.cpp]: prob make a helper function for this chunk

File: Source/Core/Core/HW/EXI_DeviceSlippi.cpp [hunk 4]
```diff
@@ -812,3 +882,214 @@ bool CEXISlippi::IsPresent() const
...
+			while (closestStateFrame > latestFrame)
+			{
+				closestStateFrame -= FRAME_INTERVAL;
+			}
+
+			INFO_LOG(SLIPPI, "Seek started, moving to frame: %d", g_targetFrameNum);
+
+			if (g_targetFrameNum < START_FRAME || closestStateFrame == START_FRAME)
+			{
+				isHardFFW = true;
```
Comment (r2dliu) [inline Source/Core/Core/HW/EXI_DeviceSlippi.cpp]: woops remove this

File: Source/Core/Common/Logging/LogManager.cpp [hunk 5]
```diff
@@ -57,6 +57,7 @@ LogManager::LogManager()
 	m_Log[LogTypes::DVDINTERFACE] = new LogContainer("DVD", "DVD Interface");
 	m_Log[LogTypes::DYNA_REC] = new LogContainer("JIT", "Dynamic Recompiler");
 	m_Log[LogTypes::EXPANSIONINTERFACE] = new LogContainer("EXI", "Expansion Interface");
+	m_Log[LogTypes::SLIPPI] = new LogContainer("SLIPPI", "Slippi");
```
Comment (r2dliu) [inline Source/Core/Common/Logging/LogManager.cpp:60]: expansioninterface was too noisy for me

File: Source/Core/Core/HW/EXI_DeviceSlippi.cpp [hunk 6]
```diff
@@ -812,3 +882,214 @@ bool CEXISlippi::IsPresent() const
...
+			isHardFFW = true;
+
+			Core::PauseAndLock(false, pause);
+			// ffw to our targetFrame
+			while (currentPlaybackFrame != g_targetFrameNum)
+			{
+				condVar.wait(lock);
+			}
+
+			pause = Core::PauseAndLock(true);
```
Comment (JLaferri) [inline Source/Core/Core/HW/EXI_DeviceSlippi.cpp]: Use emulation pause to allow frame advance

File: Source/Core/Core/HotkeyManager.cpp [hunk 7]
```diff
@@ -394,6 +400,21 @@ void HotkeyManager::LoadDefaults(const ControllerInterface& ciface)
 
     set_key_expression(HK_SHOW_OSD_CHAT, "Y");
 
+// Slippi
+#ifdef _WIN32
+	set_key_expression(HK_JUMP_BACK, "LEFT");
```
Comment (JLaferri) [inline Source/Core/Core/HotkeyManager.cpp:412]: Would be nice to have UI for hotkey stuff

**Comment by JLaferri** [comment]
1) Skipping forward/back sometimes crashes.
2) Dealing with edge cases, rewinding past start of game shouldn't be possible, should just go to first frame
3) Being able to rewind after game has ended

**Pull request merged**